pyccel/parser/semantic.py

```diff
@@ -682,154 +678,54 @@ def _infere_type(self, expr, **settings):
-                d_vars = [self._infere_type(arg,**settings) for arg in expr.args]
-                i = 0 if (d_vars[0]['rank'] >= d_vars[1]['rank']) else 1
-
-                d_var['datatype'   ] = d_vars[i]['datatype']
-                d_var['rank'       ] = d_vars[i]['rank']
-                d_var['shape'      ] = d_vars[i]['shape']
-                d_var['allocatable'] = d_vars[i]['allocatable']
-                d_var['is_pointer' ] = False
-                d_var['precision'  ] = d_vars[i].pop('precision',4)
-
-            elif name in ['Norm']:
-                d_var = self._infere_type(expr.arg,**settings)
-
-                d_var['shape'] = expr.shape(d_var['shape'])
-                d_var['rank' ] = len(d_var['shape'])
-                d_var['allocatable'] = d_var['rank']>0
-                d_var['is_pointer' ] = False
-
-            elif isinstance(expr, NumpyUfuncBase):
-                d_var = self._infere_type(expr.args[0], **settings)
-                d_var['datatype'   ] = expr.dtype
-                d_var['is_pointer' ] = False
-                d_var['precision'  ] = expr.precision
-
-            elif isinstance(expr, MathFunctionBase):
-                d_var['datatype' ] = expr.dtype
-                d_var['precision'] = expr.precision
-                d_var['rank'     ] = expr.rank
-                d_var['shape'    ] = expr.shape
-
-            elif name in ['EmptyLike', 'ZerosLike', 'OnesLike', 'FullLike']:
-                d_var = self._infere_type(expr.rhs, **settings)
+            d_var['datatype'   ] = d_vars[0]['datatype']
+            if var0_is_vector or var1_is_vector:
+                d_var['rank'   ] = 1
             else:
-                raise NotImplementedError('Type of Application : '+type(expr).__name__+' cannot be infered')
-
+                d_var['rank'   ] = 2
+            d_var['allocatable'] = False
+            d_var['precision'  ] = max(d_vars[0]['precision'],
+                                       d_vars[1]['precision'])
             return d_var
 
-        elif isinstance(expr, GC):
-            return self._infere_type(expr.lhs, **settings)
-        elif isinstance(expr, (Expr, FunctionCall)):
-            # ...
+        elif isinstance(expr ,(Array, Full, Len, Rand, NumpySum, Product,
+                               NumpyMin, NumpyMax, PythonInt, NumpyMod,
+                               PythonFloat, PythonComplex,
+                               NumpyInt, Int32, Int64,
+                               NumpyFloat, Float32, Float64,
+                               NumpyComplex, Complex64, Complex128,
+                               Real, Imag, PythonBool, Norm,
+                               NumpyUfuncBase, MathFunctionBase, Expr, FunctionCall,
+                               Integer, Float, Complex, String,
+                               BooleanTrue, BooleanFalse,
+                               PyccelArraySize, Is, IndexedElement)):
```

Comment:
Can these be grouped? E.g. `isinstance(expr, PyccelAstNode)` ?

Reply:
In theory yes, but we need more verification to be sure

Reply:
What do we need for verification? Can we not just try and see if any of the tests fail?

Reply:
I agree, it's worth to give it a try

Reply:
the  main problem with that is almost all our nodes inherit from it which means that we can never go to the else condition where we raise an error

Reply:
@saidctb What do you mean? I can see no `else` condition here, and no error raised.

Reply:
there is an else condition in the end of the method where we raise an error

Reply:
Sorry, you meant this of course:
```
        else:
            msg = '{expr} not yet available'.format(expr=type(expr))
            raise NotImplementedError(msg)
```

Reply:
So you are worried about keys other than:
- dtype
- precision
- rank
- shape
- order
not being correctly set for classes that could be added later?
(I assume no classes are reaching the `else` condition at the moment as the tests are passing)

Reply:
yes

Reply:
yes @EmilyBourne  we shouldn't go to else unless there is an error somewhere which means it's a bug and we don't test bugs , so if we add the `PyccelAstNode`  it is the same as deleting the else condition because we are always gonna pass by the condition `isinstance(expr, PyccelAstNode)`

Reply:
OK let's keep this as it is for now.

Reply:
We would go to `else` if we are accidentally looking at a sympy object or if the class has accidentally been created without inheriting from `PyccelAstNode`.
Effectively this list is there just so the error can act as a reminder to check if new classes are allocatable, etc.
It feels wrong to have 11 lines instead of 1, effectively throwing away the advantage of `PyccelAstNode`, but I can't think of a better way of doing this and it is important to ensure that these values are set when necessary

Reply:
okay I will try to make it work

Reply:
done